### PR TITLE
[SPARK-33530][CORE] Support --archives and spark.archives option natively

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

TL;DR:
- This PR completes the support of archives in Spark itself instead of Yarn-only
  - It makes `--archives` option work in other cluster modes too and adds `spark.archives` configuration.
-  After this PR, PySpark users can leverage Conda to ship Python packages together as below:
    ```python
    conda create -y -n pyspark_env -c conda-forge pyarrow==2.0.0 pandas==1.1.4 conda-pack==0.5.0
    conda activate pyspark_env
    conda pack -f -o pyspark_env.tar.gz
    PYSPARK_DRIVER_PYTHON=python PYSPARK_PYTHON=./environment/bin/python pyspark --archives pyspark_env.tar.gz#environment
   ```
- Issue a warning that undocumented and hidden behavior of partial archive handling in `spark.files` / `SparkContext.addFile` will be deprecated, and users can use `spark.archives` and `SparkContext.addArchive`.

This PR proposes to add Spark's native `--archives` in Spark submit, and `spark.archives` configuration. Currently, both are supported only in Yarn mode:

```bash
./bin/spark-submit --help
```

```
Options:
...
 Spark on YARN only:
  --queue QUEUE_NAME          The YARN queue to submit to (Default: "default").
  --archives ARCHIVES         Comma separated list of archives to be extracted into the
                              working directory of each executor.
```

This `archives` feature is useful often when you have to ship a directory and unpack into executors. One example is native libraries to use e.g. JNI. Another example is to ship Python packages together by Conda environment.

Especially for Conda, PySpark currently does not have a nice way to ship a package that works in general, please see also https://hyukjin-spark.readthedocs.io/en/stable/user_guide/python_packaging.html#using-zipped-virtual-environment (PySpark new documentation demo for 3.1.0).

The neatest way is arguably to use Conda environment by shipping zipped Conda environment but this is currently dependent on this archive feature. NOTE that we are able to use `spark.files` by relying on its undocumented behaviour that untars `tar.gz` but I don't think we should document such ways and promote people to more rely on it.

Also, note that this PR does not target to add the feature parity of `spark.files.overwrite`, `spark.files.useFetchCache`, etc. yet. I documented that this is an experimental feature as well.

### Why are the changes needed?

To complete the feature parity, and to provide a better support of shipping Python libraries together with Conda env.

### Does this PR introduce _any_ user-facing change?

Yes, this makes `--archives` works in Spark instead of Yarn-only, and adds a new configuration `spark.archives`.

### How was this patch tested?

I added unittests. Also, manually tested in standalone cluster, local-cluster, and local modes.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1537,8 +1575,14 @@ class SparkContext(config: SparkConf) extends Logging {
...
 
-  private def addFile(path: String, recursive: Boolean, addedOnSubmit: Boolean): Unit = {
-    val uri = new Path(path).toUri
+  private def addFile(
+      path: String, recursive: Boolean, addedOnSubmit: Boolean, isArchive: Boolean = false
+    ): Unit = {
+    val uri = if (!isArchive) {
+      new Path(path).toUri
+    } else {
+      Utils.resolveURI(path)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1584]: Here we cannot rely on `new Path(path).toUri`. it makes the fragment (`#`) in URI as the part of path. `Utils.resolveURI` is used for `spark.yarn.dist.archives` as well.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -1550,7 +1594,7 @@ class SparkContext(config: SparkConf) extends Logging {
 
     val hadoopPath = new Path(schemeCorrectedURI)
     val scheme = schemeCorrectedURI.getScheme
-    if (!Array("http", "https", "ftp").contains(scheme)) {
+    if (!Array("http", "https", "ftp").contains(scheme) && !isArchive) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1597]: Archive is not supposed to be a directory.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -1568,21 +1612,44 @@ class SparkContext(config: SparkConf) extends Logging {
 
     val key = if (!isLocal && scheme == "file") {
       env.rpcEnv.fileServer.addFile(new File(uri.getPath))
+    } else if (uri.getScheme == null) {
+      schemeCorrectedURI.toString
+    } else if (isArchive) {
+      uri.toString
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1618]: For the same reason of keeping the fragment, it uses URI when it's archive.

**Comment by HyukjinKwon** [comment]
@tgravescs, @mridulm, @Ngone51, can you take a look when you guys find some time?

**Comment by HyukjinKwon** [comment]
cc @zero323 and @fhoering too FYI. This is related to the docs and shipping 3rd party Python packages in PySpark apps.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 4]
```diff
@@ -535,13 +538,23 @@ private[spark] object Utils extends Logging {
...
-      executeAndGetOutput(Seq("tar", "-xf", fileName), targetDir)
+    if (shouldUntar) {
+      // Decompress the file if it's a .tar or .tar.gz
+      if (fileName.endsWith(".tar.gz") || fileName.endsWith(".tgz")) {
+        logWarning(
+          "Untarring behavior is deprecated at spark.files and " +
+            "SparkContext.addFile. Use spark.archives or SparkContext.addArchive " +
+            "instead.")
+        logInfo("Untarring " + fileName)
+        executeAndGetOutput(Seq("tar", "-xzf", fileName), targetDir)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:549]: Our `spark.files` and `SparkContext.addFile` have a sort of undocumented and hidden behaviour. Only in executor side, it untars if the files are `.tar.gz` or `tgz`. I think it makes sense to deprecate this behaviour and encourage users to use explicit archive handling.

Also, I believe it's a good practice to avoid relying on external programs anyway.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 5]
```diff
@@ -1520,6 +1534,30 @@ class SparkContext(config: SparkConf) extends Logging {
    */
   def listFiles(): Seq[String] = addedFiles.keySet.toSeq
 
+  /**
+   * Add an archive to be downloaded and unpacked with this Spark job on every node.
+   *
+   * If an archive is added during execution, it will not be available until the next TaskSet
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1540]: Is this limitation applied to the other ways like `spark.files` too?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 6]
```diff
@@ -1803,6 +1803,16 @@ package object config {
     .toSequence
     .createWithDefault(Nil)
 
+  private[spark] val ARCHIVES = ConfigBuilder("spark.archives")
+    .version("3.1.0")
+    .doc("Comma separated list of archives to be extracted into the working directory of each " +
+      "executor. .jar, .tar.gz, .tgz and .zip are supported. You can specify the directory " +
+      "name to unpack via adding '#' after the file name to unpack, for example, " +
+      "'file.zip#directory'. This configuration is experimental.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1811]: Ur, if this is still experimental, we should not deprecate the old ways.
> This configuration is experimental.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 7]
```diff
@@ -160,6 +160,85 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
     }
   }
 
+  test("basic case for addArchive and listArchives") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:163]: Could you prepend `SPARK-33530: ` please?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for generalizing this, @HyukjinKwon . It's great. I left a few comments.

One more question. Can we remove `spark.yarn.dist.archives` by making it the alternative of `spark.archives`?

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 5 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1540]: Yeah, it does. It was actually copied from `addFile`.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 6 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1811]: Makes sense. I will instead say the behaviour will be deprecated. We can remove that back once this configuration becomes stable.

**Comment by HyukjinKwon** [comment]
I think it's fine to avoid removing `spark.yarn.dist.archives` out yet - maybe we could think about removing out once this feature becomes stable (?). Yarn also has `spark.yarn.dist.files` and `spark.files` can work together as far as I know.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 8]
```diff
@@ -1568,21 +1613,38 @@ class SparkContext(config: SparkConf) extends Logging {
...
+    } else if (
+      isArchive &&
+        addedArchives.putIfAbsent(
+          UriBuilder.fromUri(new URI(key)).fragment(uri.getFragment).build().toString,
+          timestamp).isEmpty) {
+      logInfo(s"Added archive $path at $key with timestamp $timestamp")
+      val uriToDownload = UriBuilder.fromUri(new URI(key)).fragment(null).build()
+      val source = Utils.fetchFile(uriToDownload.toString, Utils.createTempDir(), conf,
+        env.securityManager, hadoopConfiguration, timestamp, useCache = false, shouldUntar = false)
+      logInfo(s"Unpacking an archive $path")
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1641]: How about logging source and dest, too?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 9]
```diff
@@ -512,6 +513,8 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
         |  --files FILES               Comma-separated list of files to be placed in the working
         |                              directory of each executor. File paths of these files
         |                              in executors can be accessed via SparkFiles.get(fileName).
+        |  --archives ARCHIVES         Comma separated list of archives to be extracted into the
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:516]: nit: `Comma separated` -> `Comma-separated` for consistency?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 10]
```diff
@@ -230,16 +232,17 @@ private[spark] class Executor(
   private val appStartTime = conf.getLong("spark.app.startTime", 0)
 
   // To allow users to distribute plugins and their required files
-  // specified by --jars and --files on application submission, those jars/files should be
-  // downloaded and added to the class loader via updateDependencies.
-  // This should be done before plugin initialization below
+  // specified by --jars, --files and --archives on application submission, those
+  // jars/files/archives should be downloaded and added to the class loader vi
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:236]: nit: `vi` -> `via`

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 11]
```diff
@@ -907,32 +911,49 @@ private[spark] class Executor(
...
         Utils.fetchFile(name, new File(SparkFiles.getRootDirectory()), conf,
           env.securityManager, hadoopConf, timestamp, useCache = !isLocal)
         currentFiles(name) = timestamp
       }
+      for ((name, timestamp) <- newArchives if currentArchives.getOrElse(name, -1L) < timestamp) {
+        logInfo(s"Fetching $name with timestamp $timestamp")
+        val sourceURI = new URI(name)
+        val uriToDownload = UriBuilder.fromUri(sourceURI).fragment(null).build()
+        val source = Utils.fetchFile(uriToDownload.toString, Utils.createTempDir(), conf,
+          env.securityManager, hadoopConf, timestamp, useCache = false, shouldUntar = false)
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:933]: We cannot use cache for this purpose?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 12]
```diff
@@ -1803,6 +1803,16 @@ package object config {
     .toSequence
     .createWithDefault(Nil)
 
+  private[spark] val ARCHIVES = ConfigBuilder("spark.archives")
+    .version("3.1.0")
+    .doc("Comma separated list of archives to be extracted into the working directory of each " +
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1808]: nit: `Comma separated` -> `Comma-separated` for consistency?

Comment (maropu) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1808]: btw, we don't need to describe something here about the relationship (or difference?) between `park.files` and `spark.archives`?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 13]
```diff
@@ -494,7 +495,8 @@ private[spark] object Utils extends Logging {
       securityMgr: SecurityManager,
       hadoopConf: Configuration,
       timestamp: Long,
-      useCache: Boolean): File = {
+      useCache: Boolean,
+      shouldUntar: Boolean = true): File = {
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:499]: In L487, could you document what the new param `shouldUntar` is?

File: docs/configuration.md [hunk 14]
```diff
@@ -784,6 +784,17 @@ Apart from these, the following properties are also available, and may be useful
   </td>
   <td>2.3.0</td>
 </tr>
+<tr>
+  <td><code>spark.archives</code></td>
+  <td></td>
+  <td>
+    Comma separated list of archives to be extracted into the working directory of each executor.
```
Comment (maropu) [inline docs/configuration.md:791]: ditto: `Comma separated` -> `Comma-separated`

**Review by maropu — COMMENTED** [review]
Nice feature! I left minor comments.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 12 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1808]: I think it's fine. They are meant to be separate configurations to end-users. These were not documented as well at least for `spark.files`, `spark.jars`, `spark.yarn.dist.files`, `spark.yarn.dist.archive`, `spark.yarn.dist.jars`, etc.

**Comment by HyukjinKwon** [comment]
Thanks @maropu and @dongjoon-hyun. I believe I addressed the comments.

**Comment by HyukjinKwon** [comment]
@mcg1969 too FYI [conda-pack](https://conda.github.io/conda-pack/). With this change, users can use conda-pack in other cluster modes not only Yarn.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
+1, LGTM (Pending CI)

**Comment by HyukjinKwon** [comment]
Thank you @dongjoon-hyun!

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
The R failure is a flaky one.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 15]
```diff
@@ -1568,21 +1612,39 @@ class SparkContext(config: SparkConf) extends Logging {
...
         env.securityManager, hadoopConfiguration, timestamp, useCache = false)
       postEnvironmentUpdate()
+    } else if (
+      isArchive &&
+        addedArchives.putIfAbsent(
+          UriBuilder.fromUri(new URI(key)).fragment(uri.getFragment).build().toString,
+          timestamp).isEmpty) {
+      logInfo(s"Added archive $path at $key with timestamp $timestamp")
+      val uriToDownload = UriBuilder.fromUri(new URI(key)).fragment(null).build()
+      val source = Utils.fetchFile(uriToDownload.toString, Utils.createTempDir(), conf,
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:1638]: > --archives ARCHIVES         Comma-separated list of archives to be extracted into the working directory of each executor.

Do we need to download archives at driver as it only declares they are extracted to executors?

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1638]: I think it's fine to unpack/download on the driver side as well to mimic the previous Yarn's behaviour. At least seems that's what Yarn does in Yarn cluster mode.

**Comment by HyukjinKwon** [comment]
I pushed some more changes to fix some nits which are all virtually non-code change (https://github.com/apache/spark/pull/30486/commits/5b1d1c3db6c1ce68a2738dec9b54af519b026b42).

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 15 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:1642]: In fact, `--files` describes same:

```
        |  --files FILES               Comma-separated list of files to be placed in the working
        |                              directory of each executor. File paths of these files
        |                              in executors can be accessed via SparkFiles.get(fileName).
```

I think we should fix the docs to say it's also available on the driver side :-). but I would like to run it separately.

**Review by maropu — APPROVED** [review]

**Comment by mridulm** [comment]
Thanks for working on this @HyukjinKwon !
I have not taken a very detailed look, but wanted to understand the interaction with use of distributed cache in yarn.
How does this coexist with that ? Or will it end up causing issues ? (archives coming in via yarn and spark executor ?)

**Comment by HyukjinKwon** [comment]
It will be exactly same as `spark.files` and `spark.yarn.dist.files`. To be honest, I am not exactly sure how they will conflictto each other but both work together as far as I know.

**Comment by HyukjinKwon** [comment]
@mridulm dose it make sense? Ill go ahead if there are not other comments :-).

**Comment by mridulm** [comment]
Thanks for the details @HyukjinKwon !
It would be good if @tgravescs can also take a look, I am a bit rusty on some of these now - and iirc there are a bunch of corner cases in this.

So my understanding is that the PR is adding functionality which existing in yarn for archives into a general support for other cluster modes too - with spark on yarn continuing to rely on distributed cache (and there should be no functionality change in yarn mode).

**Comment by HyukjinKwon** [comment]
Yeah, that's correct. One thing js though, if there's anything wrong in terms of conflict between Yarn distributed cach (spark.yarn.dist.* vs spark.* like spark.files), I would say this is a separate issue to handle since I am reusing the existing code path

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 16]
```diff
@@ -335,6 +335,44 @@ class SparkSubmitSuite
...
+    conf.get("spark.executor.cores") should be ("5")
+    conf.get("spark.jars") should include regex (".*one.jar,.*two.jar,.*three.jar")
+    conf.get("spark.files") should include regex (".*file1.txt,.*file2.txt")
+    conf.get("spark.archives") should include regex (".*archive1.zip,.*archive2.jar")
+    conf.get("spark.app.name") should be ("beauty")
+    conf.get(UI_ENABLED) should be (false)
+    sys.props("SPARK_SUBMIT") should be ("true")
+  }
+
+
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:375]: nit: redundant line

**Review by Ngone51 — COMMENTED** [review]
LGTM after taking another look.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 17]
```diff
@@ -160,6 +160,85 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
 
+  test("SPARK-33530: basic case for addArchive and listArchives") {
+    withTempDir { dir =>
+      val file1 = File.createTempFile("someprefix1", "somesuffix1", dir)
+      val file2 = File.createTempFile("someprefix2", "somesuffix2", dir)
+      val file3 = File.createTempFile("someprefix3", "somesuffix3", dir)
+      val file4 = File.createTempFile("someprefix4", "somesuffix4", dir)
+
+      val jarFile = new File(dir, "test!@$jar.jar")
+      val zipFile = new File(dir, "test-zip.zip")
```
Comment (tgravescs) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:171]: are we relying on existing test for tar.gz and tar?

**Review by tgravescs — COMMENTED** [review]
yeah the main thing is yarn still uses the distributed cache and not the spark deploy mechanism, just like it does for the files now. Doing a quick look I think this is fine.

Did you manually test it on k8s and yarn at all? that would be nice to make sure nothing unexpected.

**Comment by HyukjinKwon** [comment]
I haven't tested in K8S yet it would take me a while. I plan to add an integration test though.

Hope I can proceed it separately given that the code freeze is coming and I would like to get this in for Spark 3.1.0.

**Comment by tgravescs** [comment]
yeah its not a blocker

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 17 above)
Comment (HyukjinKwon) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala:171]: I manually tested tar.gz and tgz cases, and the cases without extensions too. The unpack method is a copy from Hadoop so I did not exhaustively test but I can add if that looks better to add the case.

**Comment by HyukjinKwon** [comment]
Thanks all @dongjoon-hyun @maropu @Ngone51 @mridulm and @tgravescs. Let me merge this in.
I will try to have some time to prepare an IT test with K8S which hopefully will be added before Spark 3.1.0 release.

Merged to master.

**Comment by HyukjinKwon** [comment]
Oh, maybe I will use tar.gz and tgz in the integration test. That will address https://github.com/apache/spark/pull/30486#discussion_r532678330 togeter.
